### PR TITLE
Event for adding modded Blocks to existing BlockEntityTypes

### PR DESCRIPTION
**Pull request by eerussianguy**
The set of valid blocks for an entity type is constructed, stored, and used in just one place, making it an easy target for modification. Modders have long been told in discord to use an AT, but it is worth adding an event giving a proper place to do the modification, and allowing only correct modification (adding blocks, rather than replacing the list or removing items, which should never happen).

One big reason for this is the `Level#getBlockEntity(pos, type)` function which only retrieves block entities of the given block entity type rather than the instanceof checks back in the 1.12 era, which forces duplicated code, headaches, etc in a lot of cases.

I'm willing to write a test mod for this -- one idea I had for this was reworking the CustomSignsTest to use this. But I'll wait to see if that's deemed necessary by triage

### CODE REVIEW

**Comment by KnightMiner** [comment]
I would suggest updating the sign test mod regardless. Modders will sometimes use forge test mods as examples of how to do a thing, and the sign test mod is the example of how to add a block to an existing block entity type (currently, by making a copy of the BE type). Will help with discoverability of this event (and will hopefully be trivial to do and help your PR)

**Comment by eerussianguy** [comment]
Had to change the set to not be an immutable set. Another option would be definalizing the variable with an AT. I think changing it to be mutable is fine because 1) it's private 2) it's typed `Set` so anyone who is already hacking into it (which we don't support anyway) is probably using it as `Set` 3) it's referenced in exactly one place

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * Provided to add new blocks to a {@link BlockEntityType}'s list of valid blocks, used for ticking, rendering, and other purposes.
+ * For your own mod's blocks, specify the blocks in the {@link BlockEntityType.Builder}, don't use this event.
+ * Blocks added to the valid blocks list are expected to have attached the same {@link BlockEntity} as the parent type in their {@link EntityBlock} implementation.
+ *
+ * Fired on the Mod bus {@link IModBusEvent}.
+ */
+public class AddValidBlocksToBlockEntityEvent extends Event implements IModBusEvent
+{
+    @ApiStatus.Internal
+    public final Map<BlockEntityType<?>, Set<Block>> newBlocks;
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: Even if marked internal, don't have public fields.

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    @ApiStatus.Internal
+    public AddValidBlocksToBlockEntityEvent()
+    {
+        newBlocks = new HashMap<>();
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: Why not do this in the field initializer? At the end of the day its the same bytecode, but looks better there.

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Adds a new {@code block} to the valid blocks list for the given {@code type}
+     */
+    public void addValidBlock(BlockEntityType<?> type, Block block)
+    {
+        if (!newBlocks.containsKey(type))
+        {
+            newBlocks.put(type, new HashSet<>());
+        }
+        newBlocks.get(type).add(block);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: I think it would be better to always use the supplier. And/Or registry name. Thus allowing registry replacements to 'just work'. This is a bit more involved patches wise as you'd be changing BlockEntityType.validBlocks to a Set<Supplier<Block>> or better yet a Set<ResourceKey<Block>>. Or do we not care about replacements?

**Review by LexManos — CHANGES_REQUESTED** [review]
Just to clarify, there are three locations in vanilla code that use this:
SignEditScreen.tick() -> Used to close the screen if the sign gets removed while open.
BlockEntityRenderDispatcher.render - > Used as a guard for all custom Block Entity Renderers to make sure the state is correct before rendering. 
LevelChunk.BoundtickingBlockEntity.tick -> Used as a guard for all block entities to prevent ticking.


This isn't a bad change, allowing reuse of vanilla BlockEntityTypes. Just needs a couple cleanup and questions answered.

File: patches/minecraft/net/minecraft/world/level/block/entity/BlockEntityType.java.patch [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    @Nullable
+@@ -115,5 +_,10 @@
+       public BlockEntityType<T> m_58966_(Type<?> p_58967_) {
+          return new BlockEntityType<>(this.f_58959_, this.f_58960_, p_58967_);
+       }
++   }
++   // FORGE START. INTERNAL USE ONLY!
++   private final java.util.Set<net.minecraft.resources.ResourceKey<Block>> additionalValidBlocks = new java.util.HashSet<>();
++   public static void fireAddBlockEvent() {
++       net.minecraftforge.common.ForgeHooks.onBlockEntityBlocks().forEach((type, blocks) -> type.additionalValidBlocks.addAll(blocks));
```
Comment (SizableShrimp) [inline patches/minecraft/net/minecraft/world/level/block/entity/BlockEntityType.java.patch]: Just because this is a patch does not mean you cannot use whitespace so long as it doesn't increase the required number of context lines. In this case, I would recommend a newline between the ending bracket and forge start comment along with a newline between the field and method.

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        final Optional<ResourceKey<Block>> resourceKey = ForgeRegistries.BLOCKS.getResourceKey(supplier.get());
+        if (resourceKey.isEmpty())
+        {
+            throw new NullPointerException("Tried to add a not registered Block to BlockEntityType: " + ForgeRegistries.BLOCK_ENTITY_TYPES.getKey(type) + " Block Class: " + supplier.get().getClass().getSimpleName());
+        }
+        if (!newBlocks.containsKey(type))
+        {
+            newBlocks.put(type, new HashSet<>());
+        }
+        newBlocks.get(type).add(resourceKey.get());
```
Comment (FiniteReality) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: From what I understand, most "register"-style events throw if the entry has been registered multiple times; is it worth doing that here?

Comment (eerussianguy) [reply src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: I think you misunderstand, multiple mods can register to the same block entity type,and mods can register muliple blocks to one block entity type, this function just ensures that a set exists to add to when one is added for the first time.

Comment (FiniteReality) [reply src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: I'm talking about the `add(resourceKey.get())` - surely each block should only have a resource key added once? ~~Would multiple mods be adding the same block to the block entity?~~ EDIT: That's not important since this is a modbus event, my bad.

Comment (FiniteReality) [reply src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: Also, in that case, wouldn't it be cleaner to do `newBlocks.computeIfAbsent(type, x -> new HashSet<>()).add(resourceKey.get())` rather than a separate contains/put?

Comment (eerussianguy) [reply src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: You are reviewing an old version of the code. Zero mentions of resource keys in the code at time of your review. But yes it could use computeIfAbsent

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Provided to add new blocks to a {@link BlockEntityType}'s list of valid blocks, used for ticking, rendering, and other purposes.
+ * For your own mod's blocks, specify the blocks in the {@link BlockEntityType.Builder}, don't use this event.
+ * Blocks added to the valid blocks list are expected to have attached the same {@link BlockEntity} as the parent type in their {@link EntityBlock} implementation.
+ *
+ * Fired on the Mod bus {@link IModBusEvent} whenever the block registry is baked. The list of forge-added valid blocks is cleared before the event is fired.
+ */
+public class AddValidBlocksToBlockEntityEvent extends Event implements IModBusEvent
+{
+    private final Map<BlockEntityType<?>, Set<Block>> newBlocks;
```
Comment (KnightMiner) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: Worth asking, does this need to be a map stored in the event? I feel like it might be better to store a `BiConsumer<BlockEntityType<?>, Block>` which just adds the block directly to the set in the block entity type. Saves the creation of additional datastructures that only exist to fill other structures.

Comment (eerussianguy) [reply src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: I guess that's just a design choice. I will leave it up to the forge team.

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 7]
```diff
@@ -28,32 +25,23 @@
  * For your own mod's blocks, specify the blocks in the {@link BlockEntityType.Builder}, don't use this event.
  * Blocks added to the valid blocks list are expected to have attached the same {@link BlockEntity} as the parent type in their {@link EntityBlock} implementation.
  *
- * Fired on the Mod bus {@link IModBusEvent}.
+ * Fired on the Mod bus {@link IModBusEvent} whenever the block registry is baked. The list of forge-added valid blocks is cleared before the event is fired.
  */
```
Comment (SizableShrimp) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java:27]: Perhaps reword to state it's cleared each time before the event is fired? And define what baking means?

File: src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java [hunk 8]
```diff
@@ -28,32 +25,23 @@
...
-        if (resourceKey.isEmpty())
-        {
-            throw new NullPointerException("Tried to add a not registered Block to BlockEntityType: " + ForgeRegistries.BLOCK_ENTITY_TYPES.getKey(type) + " Block Class: " + supplier.get().getClass().getSimpleName());
-        }
-        if (!newBlocks.containsKey(type))
-        {
-            newBlocks.put(type, new HashSet<>());
-        }
-        newBlocks.get(type).add(resourceKey.get());
+        newBlocks.computeIfAbsent(type, x -> new HashSet<>()).add(block.get());
```
Comment (SizableShrimp) [inline src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java]: Bit unnecessary to use a supplier if you immediately call get()

**Comment by eerussianguy** [comment]
Let me know if this needs any updates

File: patches/minecraft/net/minecraft/commands/CommandSourceStack.java.patch [hunk 9]
```diff
@@ -8,4 +8,4 @@
 +public class CommandSourceStack implements SharedSuggestionProvider, net.minecraftforge.common.extensions.IForgeCommandSourceStack {
     public static final SimpleCommandExceptionType f_81286_ = new SimpleCommandExceptionType(Component.m_237115_("permissions.requires.player"));
     public static final SimpleCommandExceptionType f_81287_ = new SimpleCommandExceptionType(Component.m_237115_("permissions.requires.entity"));
-    private final CommandSource f_81288_;
+    public final CommandSource f_81288_;
```
Comment (Shadows-of-Fire) [inline patches/minecraft/net/minecraft/commands/CommandSourceStack.java.patch]: This change seems unrelated and should be reverted

Comment (eerussianguy) [reply patches/minecraft/net/minecraft/commands/CommandSourceStack.java.patch]: paging @sciwhiz12  on this, as this was something I was told to leave originally and I believe it was fixed at some point, not sure though.

Comment (eerussianguy) [reply patches/minecraft/net/minecraft/commands/CommandSourceStack.java.patch]: Fixed in #9050
